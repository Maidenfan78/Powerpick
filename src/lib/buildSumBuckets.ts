export interface Bucket {
  label: string;
  mid: number;
  freq: number;
}

export function buildSumBuckets(draws: number[][], binSize = 5): Bucket[] {
  if (!draws.length) return [];

  const sums = draws.map((d) => d.reduce((a, b) => a + b, 0));
  const min = Math.min(...sums);
  const max = Math.max(...sums);

  const firstBinStart = Math.floor(min / binSize) * binSize;
  const lastBinStart = Math.ceil(max / binSize) * binSize;

  const bucketStarts: number[] = [];
  for (let s = firstBinStart; s <= lastBinStart; s += binSize) {
    bucketStarts.push(s);
  }

  const freqs = Array(bucketStarts.length).fill(0);
  sums.forEach((sum) => {
    const idx = Math.floor((sum - firstBinStart) / binSize);
    freqs[idx] += 1;
  });

  return bucketStarts.map((start, i) => ({
    label: `${start}-${start + binSize - 1}`,
    mid: start + binSize / 2,
    freq: freqs[i],
  }));
}
