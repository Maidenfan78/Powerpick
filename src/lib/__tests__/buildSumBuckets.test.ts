import { buildSumBuckets } from "../buildSumBuckets";

test("builds correct bucket frequencies", () => {
  const draws = [
    [1, 2, 3, 4, 5, 6, 7],
    [35, 34, 33, 32, 31, 30, 29],
  ];
  const buckets = buildSumBuckets(draws, 10);
  expect(buckets).toMatchSnapshot();
});
