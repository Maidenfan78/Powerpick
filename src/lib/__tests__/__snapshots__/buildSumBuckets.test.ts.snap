// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds correct bucket frequencies 1`] = `
[
  {
    "freq": 1,
    "label": "20-29",
    "mid": 25,
  },
  {
    "freq": 0,
    "label": "30-39",
    "mid": 35,
  },
  {
    "freq": 0,
    "label": "40-49",
    "mid": 45,
  },
  {
    "freq": 0,
    "label": "50-59",
    "mid": 55,
  },
  {
    "freq": 0,
    "label": "60-69",
    "mid": 65,
  },
  {
    "freq": 0,
    "label": "70-79",
    "mid": 75,
  },
  {
    "freq": 0,
    "label": "80-89",
    "mid": 85,
  },
  {
    "freq": 0,
    "label": "90-99",
    "mid": 95,
  },
  {
    "freq": 0,
    "label": "100-109",
    "mid": 105,
  },
  {
    "freq": 0,
    "label": "110-119",
    "mid": 115,
  },
  {
    "freq": 0,
    "label": "120-129",
    "mid": 125,
  },
  {
    "freq": 0,
    "label": "130-139",
    "mid": 135,
  },
  {
    "freq": 0,
    "label": "140-149",
    "mid": 145,
  },
  {
    "freq": 0,
    "label": "150-159",
    "mid": 155,
  },
  {
    "freq": 0,
    "label": "160-169",
    "mid": 165,
  },
  {
    "freq": 0,
    "label": "170-179",
    "mid": 175,
  },
  {
    "freq": 0,
    "label": "180-189",
    "mid": 185,
  },
  {
    "freq": 0,
    "label": "190-199",
    "mid": 195,
  },
  {
    "freq": 0,
    "label": "200-209",
    "mid": 205,
  },
  {
    "freq": 0,
    "label": "210-219",
    "mid": 215,
  },
  {
    "freq": 1,
    "label": "220-229",
    "mid": 225,
  },
  {
    "freq": 0,
    "label": "230-239",
    "mid": 235,
  },
]
`;
